x-pack/plugins/apm/server/saved_objects/apm_indices_space.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SavedObjectsType } from 'src/core/server';
+import { i18n } from '@kbn/i18n';
+import { ApmIndicesConfigName } from '..';
+
+const properties: { [Property in ApmIndicesConfigName]: { type: 'keyword' } } =
+  {
+    sourcemap: { type: 'keyword' },
+    error: { type: 'keyword' },
+    onboarding: { type: 'keyword' },
+    span: { type: 'keyword' },
+    transaction: { type: 'keyword' },
+    metric: { type: 'keyword' },
+  };
+
+export const apmIndicesSpace: SavedObjectsType = {
+  name: 'apm-indices-space',
+  hidden: false,
+  namespaceType: 'single',
+  mappings: { properties },
+  management: {
+    importableAndExportable: true,
```

Comment:
When we were chatting about this on Slack, I didn't realize that this object type is importable/exportable, sorry about that 🙈
This complicates things a bit; what happens if a user attempts to import an old `apm-indices` saved object?

You have a few options:

1. Leave things as-is, which will lead to a bad UX. Users will be able to import an old `apm-indices` object, but it won't change how APM behaves. When Kibana restarts, it'll attempt to migrate that to the new object and result in an error log.
2. Change `apm-indices` to use `importableAndExportable: false` -- this is a breaking change, any old exported objects won't be able to be imported in 8.2+
3. Change your approach to migrate to `namespaceType: 'single'` in-place without registering a new Saved Object type. Rough example in the diff below:

```diff
diff --git a/x-pack/plugins/apm/server/routes/settings/apm_indices/get_apm_indices.ts b/x-pack/plugins/apm/server/routes/settings/apm_indices/get_apm_indices.ts
index e0b63ba6a2f..0b10a41b234 100644
--- a/x-pack/plugins/apm/server/routes/settings/apm_indices/get_apm_indices.ts
+++ b/x-pack/plugins/apm/server/routes/settings/apm_indices/get_apm_indices.ts
@@ -23,12 +23,12 @@ async function getApmIndicesSavedObject(
   savedObjectsClient: ISavedObjectsClient
 ) {
   const apmIndices = await withApmSpan('get_apm_indices_saved_object', () =>
-    savedObjectsClient.get<Partial<ApmIndicesConfig>>(
-      APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE,
-      APM_INDEX_SETTINGS_SAVED_OBJECT_ID
-    )
+    savedObjectsClient.get<
+      Partial<ApmIndicesConfig & { isSpaceAware: boolean }>
+    >(APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE, APM_INDEX_SETTINGS_SAVED_OBJECT_ID)
   );
-  return apmIndices.attributes;
+  const { isSpaceAware, ...attributes } = apmIndices.attributes;
+  return attributes;
 }
 
 export function getApmIndicesConfig(config: APMConfig): ApmIndicesConfig {
diff --git a/x-pack/plugins/apm/server/routes/settings/apm_indices/save_apm_indices.ts b/x-pack/plugins/apm/server/routes/settings/apm_indices/save_apm_indices.ts
index 316bf9910b3..f7d6d2b2858 100644
--- a/x-pack/plugins/apm/server/routes/settings/apm_indices/save_apm_indices.ts
+++ b/x-pack/plugins/apm/server/routes/settings/apm_indices/save_apm_indices.ts
@@ -20,7 +20,10 @@ export function saveApmIndices(
   return withApmSpan('save_apm_indices', () =>
     savedObjectsClient.create(
       APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE,
-      removeEmpty(apmIndices),
+      {
+        ...removeEmpty(apmIndices),
+        isSpaceAware: true,
+      },
       {
         id: APM_INDEX_SETTINGS_SAVED_OBJECT_ID,
         overwrite: true,
diff --git a/x-pack/plugins/apm/server/saved_objects/apm_indices.ts b/x-pack/plugins/apm/server/saved_objects/apm_indices.ts
index 4aa6c495305..a7e278e3f2b 100644
--- a/x-pack/plugins/apm/server/saved_objects/apm_indices.ts
+++ b/x-pack/plugins/apm/server/saved_objects/apm_indices.ts
@@ -23,7 +23,7 @@ const properties: { [Property in ApmIndicesConfigName]: { type: 'keyword' } } =
 export const apmIndices: SavedObjectsType = {
   name: 'apm-indices',
   hidden: false,
-  namespaceType: 'agnostic',
+  namespaceType: 'single',
   mappings: { properties },
   management: {
     importableAndExportable: true,
diff --git a/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts b/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts
index 201cce16547..7d11d0b9aae 100644
--- a/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts
+++ b/x-pack/plugins/apm/server/saved_objects/migrations/migrate_legacy_apm_indices_to_space_aware.ts
@@ -16,34 +16,19 @@ import {
 } from '../../../common/apm_saved_object_constants';
 import { ApmIndicesConfig } from '../../routes/settings/apm_indices/get_apm_indices';
 
-// Legacy APM indices saved objects
-const APM_INDICES_LEGACY_SAVED_OBJECT_TYPE = 'apm-indices';
-const APM_INDICES_LEGACY_SAVED_OBJECT_ID = 'apm-indices';
-
-/**
- * Legacy APM indices config object attributes (pre-8.2).
- * We need to keep a snapshot of this interface to guard against any incompatible ApmIndicesConfig interface changes in the future.
- */
-interface LegacyApmIndicesConfig {
-  sourcemap: string;
-  error: string;
-  onboarding: string;
-  span: string;
-  transaction: string;
-  metric: string;
-  apmAgentConfigurationIndex: string;
-  apmCustomLinkIndex: string;
-}
-
 async function fetchLegacyAPMIndices(repository: ISavedObjectsRepository) {
   try {
-    const legacyAPMIndices = await repository.get<
-      Partial<LegacyApmIndicesConfig>
-    >(APM_INDICES_LEGACY_SAVED_OBJECT_TYPE, APM_INDICES_LEGACY_SAVED_OBJECT_ID);
-    return legacyAPMIndices;
+    const apmIndices = await repository.get<
+      Partial<ApmIndicesConfig & { isSpaceAware: boolean }>
+    >(APM_INDEX_SETTINGS_SAVED_OBJECT_ID, APM_INDEX_SETTINGS_SAVED_OBJECT_TYPE);
+    if (apmIndices.attributes.isSpaceAware) {
+      // This has already been migrated to become space-aware
+      return null;
+    }
+    return apmIndices;
   } catch (err) {
     if (SavedObjectsErrorHelpers.isNotFoundError(err)) {
-      // This is expected after the legacy APM indices object has been migrated for the first time.
+      // This can happen if APM is not being used.
       return null;
     }
     throw err;
@@ -82,12 +67,6 @@ export async function migrateLegacyAPMIndicesToSpaceAware({
         attributes: legacyAPMIndices.attributes,
       }))
     );
-
-    // Delete legacy APM indices
-    await repository.delete(
-      APM_INDICES_LEGACY_SAVED_OBJECT_TYPE,
-      APM_INDICES_LEGACY_SAVED_OBJECT_ID
-    );
   } catch (e) {
     logger.error('Failed to migrate legacy APM indices object: ' + e.message);
   }
```

Then you should be able to allow users to seamlessly import `apm-indices` objects from old Kibana versions, and you won't have to maintain two saved object types 🎉 
I have not tested it but I think it should work.

I also looked at [SO import hooks](https://github.com/elastic/kibana/blob/ca05637afa9fc822d0e7abd6f98c6370d2327656/src/core/server/saved_objects/types.ts#L426-L468), but this only allows you to surface warnings to the user, so I don't think it's very helpful here.

Reply:
Thanks for all the help @jportner !

Reply:
changes updated here https://github.com/elastic/kibana/pull/126176/commits/5041281b067bd2d0c4da0a37b8fa13c137667148